Implement validation of the core game config WML

CelticMinstrel opened the pull request:
The schema validator has been around for who knows how long, but it was only ever used for the GUI2 WML. I decided it would be useful to extend it to work for the game data WML as well. This commit adds some additional features to the schema validator as well as creating a schema definition that covers about 90% of the core game data structure.

In other words, it's not done yet, but I'm opening it up for preliminary review.

Through this work I uncovered several errors which I will eventually push fixes for. For one example, a Dwarvish Thunderer who killed an enemy with his blade will then blow the smoke off of his firearm, which is not the intent of that animation.

gfgtdf commented on data/schema/core/addons.cfg:
```diff
@@ -0,0 +1,470 @@
+	{SIMPLE_KEY except string_list}
+	{SIMPLE_KEY terrain_liked terrain_list}
+
+	{DEFAULT_KEY allow_player bool yes}
```
i'm quite sure this and below are not allowed in [multiplayer faction]

CelticMinstrel replied:
I don't recall why I included then there, but... according to the wiki, `[multiplayer_side]` accepts everything that `[side]` does. In fact, it says the former is simply merged into the latter when a game starts.

CelticMinstrel replied:
Though thinking further on it, even if it's allowed, `allow_player` at the very least would completely defeat the point of the `[multiplayer_side]` tag.

CelticMinstrel replied:
I think what happened here is I cut-and-pasted these from another area but intended to drop them in `[side]`, not `[multiplayer_side]`.

gfgtdf commented on data/schema/core/addons.cfg:
```diff
@@ -0,0 +1,470 @@
+	max=infinite
+	super="$addon"
+	{ADDON_DESCRIPTION}
+	# TODO: Is this the correct default value?
```
iirc it's hybrid, not suere though

gfgtdf commented on data/schema/core/addons.cfg:
```diff
@@ -0,0 +1,470 @@
+	{DEFAULT_KEY experience_modifier int_percent 100}
+	{DEFAULT_KEY current_time int 0}
+	{SIMPLE_KEY map_generation string}
+	{SIMPLE_KEY scenario_generation string}
```
only one of map_generation /scenario_generation  can be used not both at the same time.

CelticMinstrel replied:
The schema currents is incapable of representing that restraint. Not sure if it's reply that important to be honest.

gfgtdf commented on data/schema/core/addons.cfg:
```diff
@@ -0,0 +1,470 @@
+		max=infinite
+		super="$side"
+		{REQUIRED_KEY side int}
+		# TODO: Pretty sure this isn't actually required?
```
technicially not, but it has a different default value in sp and mp, so if you want your campaign to work in both in sp and mp you should specify it, (in particular since iirc 1.13 allow syou to start sp campaign in the mp create dialog). We shodul really make this key manditory.

CelticMinstrel replied:
I think making it mandatory might be a bit much…

CelticMinstrel commented:
One weird error I'm getting is "Tag [event] may not be used in [titlescreen_music]". Why is there an `[event]` tag there? Is it really allowed, or is it ignored?

CelticMinstrel commented:
Alright, this covers the main game and editor config. The only other issue found was a `turns=unlimited` in A New Land, which I haven't fixed mainly because I'm unsure whether it's actually wrong.

The final commit (6153887) does not depend on the schema system and can be cherry-picked.

gfgtdf commented:
since this feature is rally more of an external tool than doesn't add any compability issues there is no reason why we cannot put it in 1.14.1

CelticMinstrel commented:
That last commit only addresses the issues already found, excluding campaigns (but including the tutorial).

The schema itself is not finished yet, so what I intend to do is finish it and then rerun it to see if there are further issues. After that I'll also run it on each campaign. I expect it's likely that there will be additional fix commits.

By the way, does anyone know the answer regarding the music events?

CelticMinstrel commented:
(Oh right @gfgtdf – this is not an external tool. It's something built interesting the engine which has already been in use for the GUI2 WML. No idea whether that affects its suitability for 1.14 though. I'll let you and @jyrkive argue about that.)

jyrkive commented:
@gfgtdf Fine for me. As long as the validation is disabled by default and 1.14.1 is not an urgent release, I'm okay with merging this validation for 1.14.1.

gfgtdf commented:
i know that this is a part of wesnoth.exe, but it is something that
1) is not activated by default (afaik unlike the gui2 case where the validator always runs ? not sure here) and should  have no effect unless you explicitly enable this 
2) It doesnt cause any (forward or backwards) compabiltiy issues, (in particular regarding umc)

The only thing that really could go wrong here is that the code changes somehow break the gui2 validaiton, assuming that umc gui2 (wesnoth.show_dialog etc.) is not run thought the validator and theat users don't significantly change the core gui2wml files, the fact that it didn't gave us false positive results in the gui2 yet, is probabyl good enough to assume that it won't casue bad errors there.

jyrkive commented:
Indeed, GUI2 WML is validated by default. I have personal experience about that when messing with GUI2.

CelticMinstrel commented:
Yeah, it's activated by a command-line flag. Of course I'd make sure the GUI2 WML still validates, and user-defined dialogs are generally done in Lua so there's no way to validate that.

gfgtdf commented:
if you correct your branch tey'll propbably disappear automaticially, currently the pr goes against master but was obviously branched off from 1.14

CelticMinstrel commented:
That's not an option right now. I'm purposely working on 1.14 for the time being since the goal is to fix any issues found for 1.14.

gfgtdf commented:
please make it target 1.14. It's impossible to get an overview of this othwewise.

CelticMinstrel commented:
This is not going to be made to target 1.14 and will never be merged for 1.14. It's not ready for review either. Once it _is_ ready for review, I'll rebase it against master so you can more easily see the changes.

CelticMinstrel commented:
In case someone intends to look at it, as of the time of this posting the first actual commit in the PR is https://github.com/wesnoth/wesnoth/pull/2562/commits/f64f242f00d514029a595f8fa613b1abb7564a40.

Eventually this will be rebased against master though, making it a lot easier to see (and stripping out several irrelevant commits that have already been cherry-picked to master).

CelticMinstrel commented:
This is ready for review with the intent of merging &ndash; the validator work is finished, though I might want to make more changes to the command-line interface for invoking the validator, and maybe a little cleanup of the actual schema files.

gfgtdf commented:
i still don't like the adding the redundand side= in [side][unit] change, is it really not possible to tell the validator that the side= is not needed there ?

CelticMinstrel commented:
For people feeling daunted by the huge size of this review, my suggestion is to either:

- Focus only on the schema itself (data/schema/game_config.cfg and files included from there) and comment on how closely it conforms to documented WML (note that by now it's probably a little out-of-date with respect to newly-added WML stuff; if anyone spots something like that, do let me know so I can fix it).
- Focus only on the C++ source code changes, with particular emphasis on the command-line arguments.

> i still don't like the adding the redundand side= in [side][unit] change, is it really not possible to tell the validator that the side= is not needed there ?

It's possible to tell the validator that the `side=` is not needed there, and indeed if you have a `[unit]` in `[side]` that does not have `side=`, the validator already correctly accepts it as valid (if I recall correctly).

The problem here is that the macro adds an empty `side=` key. Empty is not a valid value for `side=`, so this causes the validator to complain. And we _don't_ want empty to be a valid value in `[unit]` outside of `[side]`; the same schema definition is used in both places.
